administrative-sdk/basicAuth.js

```diff
@@ -38,55 +38,32 @@ class BasicAuth {
   }
 
   /**
-   * Callback used by createBasicAuth.
-   *
-   * @callback Sdk~basicAuthCreatedCallback
-   * @param {its.BasicAuth} basicAuth Updated basicAuth domain model instance.
-   */
-  basicAuthCreatedCallback(basicAuth) {
-  }
-
-  /**
-   * Error callback used by createBasicAuth.
-   *
-   * @callback Sdk~basicAuthCreatedErrorCallback
-   * @param {object[]} errors Array of errors.
-   * @param {its.BasicAuth} basicAuth BasicAuth domain model instance with unapplied changes.
-   */
-  basicAuthCreatedErrorCallback(errors, basicAuth) {
-  }
-
-  /**
    * Create a basic auth.
    *
    * @param {its.BasicAuth} basicAuth A basic auth domain model instance.
    * @param {Sdk~basicAuthCreatedCallback} [cb] The callback that handles the response.
    * @param {Sdk~basicAuthCreatedErrorCallback} [ecb] The callback that handles the error response.
    */
-  createBasicAuth(connection, cb, ecb) {
+  createBasicAuth(connection) {
     var self = this;
-    var _cb = function(data) {
-      self.principal = data.principal;
-      self.created = new Date(data.created);
-      self.updated = new Date(data.updated);
-      // Credentials are only supplied when generated by the backend.
-      if (data.credentials) {
-        self.credentials = data.credentials;
-      }
-      if (cb) {
-        cb(self);
-      }
-    };
-
-    var _ecb = function(errors) {
-      if (ecb) {
-        ecb(errors, self);
-      }
-    };
-
     var url = connection.settings.apiUrl + '/basicauths';
-    var fd = JSON.stringify(this);
-    connection._secureAjaxPost(url, fd, _cb, _ecb);
+    var formData = JSON.stringify(self);
+    return new Promise(function(resolve, reject) {
+      connection._secureAjaxPost(url, formData)
+        .then(function(data) {
+          self.principal = data.principal;
+          self.created = new Date(data.created);
+          self.updated = new Date(data.updated);
+          // Credentials are only supplied when generated by the backend.
+          if (data.credentials) {
+            self.credentials = data.credentials;
+          }
+          resolve(self);
+        })
+        .catch(function(error) {
+          reject(error);
+        });
+    });
```

Comment:
One of the great advantages is to prevent callback hell. This can be done by chaining them. This piece of code could be greatly simplified

```js
return connection._secureAjaxPost(url, JSON.stringify(this))
  .then(data => {
    this.principal = data.principal;
    this.created = new Date(data.created);
    this.updated = new Date(data.updated);
    // Credentials are only supplied when generated by the backend.
    if (data.credentials) {
      this.credentials = data.credentials;
    }
    return this;
  });
```

Also note that `this` refers to the `basicAuth` instance, because an arrow function is used. There's barely any need to ever reassign `this` to `self` anymore.

This comment applies to many of the same types of refactor I've seen so far.